### PR TITLE
New test/test target teams distribute thread limit

### PR DESCRIPTION
**Pull request by jhdavis8**
This test checks the `thread_limit(integer)` clause, which binds to the `teams` construct. It does so by first running a target teams distribute with no clauses to get the default thread limit, and then running it once more with the `thread_limit(default_threads - 1)`. It fails if the actual `thread_limit` is above the specified number.

These tests pass all Summit compilers.

### CODE REVIEW

File: tests/4.5/target_teams_distribute/test_target_teams_distribute_thread_limit.c [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+int main() {
+  int isOffloading;
+  OMPVV_TEST_AND_SET_OFFLOADING(isOffloading);
+  int default_threads;
+  int num_threads;
+  int errors = 0;
+
+#pragma omp target teams distribute map(from: default_threads)
+  for (int x = 0; x < N; ++x) {
+    default_threads = omp_get_thread_limit();
```
Comment (josemonsalve2) [inline tests/4.5/target_teams_distribute/test_target_teams_distribute_thread_limit.c]: data race

File: tests/4.5/target_teams_distribute/test_target_teams_distribute_thread_limit.c [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+  }
+
+  if (default_threads == 1) {
+    OMPVV_WARNING("Test operated with one thread. Cannot test thread_limit clause.");
+  } else if(default_threads <= 0) {
+    OMPVV_ERROR("omp_get_thread_limit() returned thread_limit <= 0.");
+    errors = 1;
+  } else {
+#pragma omp target teams distribute thread_limit(default_threads - 1) map(from: default_threads)
```
Comment (josemonsalve2) [inline tests/4.5/target_teams_distribute/test_target_teams_distribute_thread_limit.c]: `default_threads/2` might be more interesting?

File: tests/4.5/target_teams_distribute/test_target_teams_distribute_thread_limit.c [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+    for (int x = 0; x < N; ++x) {
+      num_threads = omp_get_thread_limit();
```
Comment (josemonsalve2) [inline tests/4.5/target_teams_distribute/test_target_teams_distribute_thread_limit.c]: data race

File: tests/4.5/target_teams_distribute/test_target_teams_distribute_thread_limit.c [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+    }
+    if (num_threads > default_threads - 1) {
+      errors += 1;
+      OMPVV_ERROR("Test ran on more threads than requested.");
+      return errors;
```
Comment (josemonsalve2) [inline tests/4.5/target_teams_distribute/test_target_teams_distribute_thread_limit.c]: This should be `OMPVV_TEST_AND_SET`

File: tests/4.5/target_teams_distribute/test_target_teams_distribute_thread_limit.c [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+    } else if (num_threads < default_threads - 1) {
+      OMPVV_WARNING("Test ran on less threads than requested.  Still spec-conformant.");
```
Comment (josemonsalve2) [inline tests/4.5/target_teams_distribute/test_target_teams_distribute_thread_limit.c]: thread_limit does not necessarily work as num_threads, so we are not technically requesting threads, just limiting them.

File: tests/4.5/target_teams_distribute/test_target_teams_distribute_thread_limit.F90 [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+
+  OMPVV_REPORT_AND_RETURN()
+CONTAINS
+  INTEGER FUNCTION test_threads()
+    INTEGER:: num_threads, default_threads, errors, x
+    errors = 0
+    
+    !$omp target teams distribute map(from: default_threads)
+    DO x = 1, N
+       default_threads = omp_get_thread_limit()
```
Comment (josemonsalve2) [inline tests/4.5/target_teams_distribute/test_target_teams_distribute_thread_limit.F90]: data race

File: tests/4.5/target_teams_distribute/test_target_teams_distribute_thread_limit.F90 [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+    ELSE
+       !$omp target teams distribute map(from: num_threads) thread_limit(&
+       !$omp& default_threads - 1)
+       DO x = 1, N
+          num_threads = omp_get_thread_limit()
+       END DO
+
+       IF (num_threads .gt. default_threads - 1) THEN
+          errors = errors + 1
+          OMPVV_ERROR("Test ran on more threads than requested.")
```
Comment (josemonsalve2) [inline tests/4.5/target_teams_distribute/test_target_teams_distribute_thread_limit.F90]: OMPVV_TEST

File: tests/4.5/target_teams_distribute/test_target_teams_distribute_thread_limit.F90 [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+    IF (default_threads .eq. 1) THEN
+       OMPVV_WARNING("Test operated with one thread. Cannot test thread_limit.")
+    ELSE IF (default_threads .lt. 1) THEN
+       OMPVV_ERROR("Test returned thread_limit <= 0")
+       errors = errors + 1
+    ELSE
+       !$omp target teams distribute map(from: num_threads) thread_limit(&
+       !$omp& default_threads - 1)
+       DO x = 1, N
+          num_threads = omp_get_thread_limit()
```
Comment (josemonsalve2) [inline tests/4.5/target_teams_distribute/test_target_teams_distribute_thread_limit.F90]: datarace

File: tests/4.5/target_teams_distribute/test_target_teams_distribute_thread_limit.F90 [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+    END DO
+
+    IF (default_threads .eq. 1) THEN
+       OMPVV_WARNING("Test operated with one thread. Cannot test thread_limit.")
+    ELSE IF (default_threads .lt. 1) THEN
+       OMPVV_ERROR("Test returned thread_limit <= 0")
+       errors = errors + 1
+    ELSE
+       !$omp target teams distribute map(from: num_threads) thread_limit(&
+       !$omp& default_threads - 1)
```
Comment (josemonsalve2) [inline tests/4.5/target_teams_distribute/test_target_teams_distribute_thread_limit.F90]: `thread_limit()` in the same line to make it cleaner. 
I suggest `default_threads/2`

File: tests/4.5/target_teams_distribute/test_target_teams_distribute_thread_limit.F90 [hunk 10]
```diff
@@ -0,0 +1,63 @@
+!===--- test_target_teams_distribute_num_threads.F90------------------------===//
+!
+! OpenMP API Version 4.5 Nov 2015
+!
+! This test uses the num_threads clause on a target teams distribute directive to
+! indicate a requested number of threads to execute the teams distribute region.
+! The specifications indicate that the number of threads that are given can be any
```
Comment (spophale) [inline tests/4.5/target_teams_distribute/test_target_teams_distribute_thread_limit.F90]: specification indicates

File: tests/4.5/target_teams_distribute/test_target_teams_distribute_thread_limit.c [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+  if (default_threads > 0) {
+#pragma omp target teams distribute thread_limit(default_threads / 2) map(from: default_threads)
+    for (int x = 0; x < N; ++x) {
+      if (omp_get_team_num() == 0) {
+        num_threads = omp_get_thread_limit();
+      }
+    }
+
+    OMPVV_TEST_AND_SET(errors, num_threads > default_threads / 2);
+    OMPVV_WARNING_IF(num_threads < default_threads / 2, "Test limited to fewer threads than were indicated. Still spec-conformant.");
```
Comment (spophale) [inline tests/4.5/target_teams_distribute/test_target_teams_distribute_thread_limit.c]: Something more meaningful---like thread limit clause was not observed, provided fewer threads than indicated.

**Pull request merged**